data-prepper-plugins/s3-source/src/main/java/org/opensearch/dataprepper/plugins/source/s3/S3ScanPartitionCreationSupplier.java

```diff
@@ -141,32 +148,17 @@ private List<PartitionIdentifier> listFilteredS3ObjectsForBucket(final List<Stri
                     .filter(keyTimestampPair -> isKeyMatchedBetweenTimeRange(keyTimestampPair.right(), startDateTime, endDateTime, isFirstScan))
                     .map(Pair::left)
                     .map(objectKey -> PartitionIdentifier.builder().withPartitionKey(String.format(BUCKET_OBJECT_PARTITION_KEY_FORMAT, bucket, objectKey)).build())
-                    .collect(Collectors.toList()));
-
+                    .collect(Collectors.toList());
             LOG.info("Found page of {} objects from bucket {}", listObjectsV2Response.keyCount(), bucket);
-        } while (listObjectsV2Response.isTruncated());
-
-        if (folderPartitioningOptions != null) {
-            final Set<PartitionIdentifier> folderPartitions = allPartitionIdentifiers.stream()
-                    .map(partitionIdentifier -> {
-                        final String fullObjectKey = partitionIdentifier.getPartitionKey();
-                        final String prefix = getPrefixWithDepth(fullObjectKey);
-                        if (prefix == null) {
-                            return null;
-                        }
-                        return PartitionIdentifier.builder().withPartitionKey(prefix).build();
-                    })
-                    .filter(Objects::nonNull)
-                    .collect(Collectors.toSet());
-
-            LOG.info("Running in folder_partitions mode at depth {}, found {} unique prefixes from {} objects", folderPartitioningOptions.getFolderDepth(), folderPartitions.size(), allPartitionIdentifiers.size());
-
-            return new ArrayList<>(folderPartitions);
-        } else {
-            LOG.info("Returning partitions for {} S3 objects from bucket {}", allPartitionIdentifiers.size(), bucket);
-        }
 
-        return allPartitionIdentifiers;
+            if (folderPartitioningOptions != null) {
+                final List<PartitionIdentifier> folderPartitionsForPage = getFolderPartitionIdentifiers(partitionsForPage);
+                sourceCoordinator.createPartitions(folderPartitionsForPage);
+            } else {
+                LOG.info("Creating partitions for {} S3 objects from bucket {}", partitionsForPage.size(), bucket);
+                sourceCoordinator.createPartitions(partitionsForPage);
```

Comment:
It is worth considering adding some way to update global state ownership periodically in this call to create partitions as this can potentially take a long time. I would probably have the LeaseBasedSourceCoordinator do this within this method.